### PR TITLE
Created docker files for an integ test cluster (#601)

### PR DESCRIPTION
**Pull request by normanj-bitquill**
### Description
Created a cluster that can later be used for integration tests. It contains a `docker-compose.yml` file that can be used to start the whole cluster.

Cluster contains:
* Spark master
* Spark worker
* OpenSearch server
* OpenSearch dashboards
* Minio server

Currently the Minio server is unused.

Spark nodes are configured to include the Flint and PPL extensions as well as to be able to query the OpenSearch server.

The OpenSearch dashboards are configured to connect to the OpenSearch server.

### Related Issues
https://github.com/opensearch-project/opensearch-spark/issues/601

### Check List
- [ ] Updated documentation (docs/ppl-lang/README.md)
- [ ] Implemented unit tests
- [ ] Implemented tests for combination with other commands
- [ ] New added source code should include a copyright header
- [ ] Commits are signed per the DCO using `--signoff`

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/sql/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by normanj-bitquill** [comment]
@YANG-DB I am part way through altering the integ test script to run against the docker containers. I have been able to create the indices for `http_logs` and `nested`. Those two indices cover about half of the tests.

Some tests now pass when they were expected to fail. This could be caused by more recent changes.

Some tests fail when they were expected to pass. These fall into 3 categories:
* Refer to a field that doesn't exist (such as `stat` from `http_logs` rather than `status`)
* Typos (such as wrong character for single quote or too many starting parenthesis)
* Query fails with no obvious explanation. Also fails when run in Spark Shell. For example: https://github.com/opensearch-project/opensearch-spark/blob/main/integ-test/script/test_cases.csv?plain=1#L21

I will continue to update the script for running the tests to also get the report at the end.

**Review by YANG-DB — COMMENTED** [review]
@normanj-bitquill how would spark-connect be used ?
will it be via python ? scala ?
could you plz describe the use case ?

**Comment by normanj-bitquill** [comment]
@YANG-DB I have been repurposing the script:
https://github.com/opensearch-project/opensearch-spark/blob/main/integ-test/script/SanityTest.py

With that it is:
Python script -> Spark Connect -> Spark Master Node

This would be an initial phase in this PR. The follow up PR would be to make use of the Scala integration test framework already in place. Update it to connect with Spark Connect and run tests.

**Comment by YANG-DB** [comment]
> @YANG-DB I have been repurposing the script:
> https://github.com/opensearch-project/opensearch-spark/blob/main/integ-test/script/SanityTest.py
> 
> With that it is:
> Python script -> Spark Connect -> Spark Master Node
> 
> This would be an initial phase in this PR. The follow up PR would be to make use of the Scala integration test framework already in place. Update it to connect with Spark Connect and run tests.

I'm not sure EMR supports spark connect...

**Comment by normanj-bitquill** [comment]
I doubt that EMR would support Spark Connect. I am keeping that in mind, but I don't have an obvious solution for Spark EMR as yet. In the end the integration tests need to be able to run queries against either standard Spark containers or Spark EMR. The integration tests should not care which they are using.

When I get to creating docker files for integration tests with Spark EMR, I will find a solution to this problem. It may require altering how integration tests connect to run queries, but for now I'd like to get a starting point out.

**Comment by normanj-bitquill** [comment]
@YANG-DB I have updated this PR so that the Python script for integration tests will now run against the docker cluster.

Below is one idea for the long term solution of running integration tests. Let me know what you think and if we should discuss this elsewhere.

## Proposal

Create a directory structure for the tests.
```
integ-test-data
  +- queries
  +- query-plans
  +- expected-results
```

`queries` - contains the queries. One query per file.
`query-plans` - expected query plans with names that correspond to filenames in `queries`
`expected-results` - expected results of the queries in `queries`, with names that correspond to filenames in `queries`

Create a Spark App that makes use of the `integ-test-data` directory. It runs each query and places the output into another directory. It also calls EXPLAIN for each query and places the output into another directory.

The Spark (either master container or EMR container) have the following directories mounted:
* `integ-test-data`
* `query-results`
* `explain-results`

The integration tests (run from `sbt`) will start the docker cluster and then upload the Spark App by either calling `spark-submit` remotely or using docker to run `spark-submit`.

After the tests finish, the integration tests (run from `sbt`) examine the query results and explain results to verify if they match the expected results.

**Comment by YANG-DB** [comment]
> @YANG-DB I have updated this PR so that the Python script for integration tests will now run against the docker cluster.
> 
> Below is one idea for the long term solution of running integration tests. Let me know what you think and if we should discuss this elsewhere.
> 
> ## Proposal
> Create a directory structure for the tests.
> 
> ```
> integ-test-data
>   +- queries
>   +- query-plans
>   +- expected-results
> ```
> 
> `queries` - contains the queries. One query per file. `query-plans` - expected query plans with names that correspond to filenames in `queries` `expected-results` - expected results of the queries in `queries`, with names that correspond to filenames in `queries`
> 
> Create a Spark App that makes use of the `integ-test-data` directory. It runs each query and places the output into another directory. It also calls EXPLAIN for each query and places the output into another directory.
> 
> The Spark (either master container or EMR container) have the following directories mounted:
> 
> * `integ-test-data`
> * `query-results`
> * `explain-results`
> 
> The integration tests (run from `sbt`) will start the docker cluster and then upload the Spark App by either calling `spark-submit` remotely or using docker to run `spark-submit`.
> 
> After the tests finish, the integration tests (run from `sbt`) examine the query results and explain results to verify if they match the expected results.

@normanj-bitquill 
Thanks for the feedback - lets take the discussion and create a dedicated issue (RFC) for that

**Comment by normanj-bitquill** [comment]
@YANG-DB I have created this issue
https://github.com/opensearch-project/opensearch-spark/issues/992

to continue discussion of how integration tests could be run on each of the Docker clusters.

**Review by YANG-DB — APPROVED** [review]
@normanj-bitquill 
looks great , can you please add some link to the `./script/README.md` file from our main `readme.md` file ?
right below [this](https://github.com/opensearch-project/opensearch-spark?tab=readme-ov-file#ppl-run-queries-on-a-local-spark-cluster)

File: integ-test/script/README.md [hunk 1]
```diff
@@ -17,21 +17,31 @@ Apart from the basic feature, it also has some advanced functionality includes:
 ### Usage
 To use this script, you need to have Python **3.6** or higher installed. It also requires the following Python libraries:
 ```shell
-pip install requests pandas openpyxl
+pip install requests pandas openpyxl pyspark setuptools pyarrow grpcio grpcio-status protobuf
+```
+
```
Comment (YANG-DB) [inline integ-test/script/README.md:22]: plz also mention that both `ppl-spark-integration-assembly-0.7.0-SNAPSHOT.jar` & `flint-spark-integration-assembly-0.7.0-SNAPSHOT.jar` needed to be build using :
- `sbt clean sparkSqlApplicationCosmetic/assembly`
- `sbt clean sparkPPLCosmetic/assembly`
before the docker can run...

File: integ-test/script/README.md [hunk 2]
```diff
@@ -17,21 +17,31 @@ Apart from the basic feature, it also has some advanced functionality includes:
...
+python SanityTest.py --base-url ${URL_ADDRESS} --username *** --password *** --opensearch-url ${OPENSEARCH_URL} --input-csv test_cases.csv --output-file test_report
 ```
-You need to replace the placeholders with your actual values of URL_ADDRESS, DATASOURCE_NAME and USERNAME, PASSWORD for authentication to your endpoint.
+You need to replace the placeholders with your actual values of URL_ADDRESS, OPENSEARCH_URL and USERNAME, PASSWORD for authentication to your endpoint.
 
 For more details of the command line parameters, you can see the help manual via command:
 ```shell
 python SanityTest.py --help   
 
 usage: SanityTest.py [-h] --base-url BASE_URL --username USERNAME --password PASSWORD --datasource DATASOURCE --input-csv INPUT_CSV
```
Comment (YANG-DB) [inline integ-test/script/README.md:43]: what is an example value for ${URL_ADDRESS} ? if it the spark's  url ?
please mention that

File: integ-test/script/README.md (hunk 1 above)
Comment (normanj-bitquill) [reply integ-test/script/README.md:22]: Added this section.

File: integ-test/script/README.md (hunk 2 above)
Comment (normanj-bitquill) [reply integ-test/script/README.md]: Fixed this up. It should actually be `SPARK_URL`. Also provided an example value.

**Comment by normanj-bitquill** [comment]
> @normanj-bitquill looks great , can you please add some link to the `./script/README.md` file from our main `readme.md` file ? right below [this](https://github.com/opensearch-project/opensearch-spark?tab=readme-ov-file#ppl-run-queries-on-a-local-spark-cluster)

Added a link in the top-level README.md

**Review by YANG-DB — APPROVED** [review]

**Comment by normanj-bitquill** [comment]
@YANG-DB I have added a section to the integ test README.md to describe the test indices.

**Review by YANG-DB — APPROVED** [review]

**Pull request merged**